### PR TITLE
Add `cleanup` option

### LINKED ISSUE #51: Kill child when exiting
If I use `execa.shell` to spawn a long running process, but the original process dies for some reason it would be great if the spawned process gets terminated automatically.

A small example

```js
'use strict'

const execa = require('execa')

execa.shell('ipfs daemon')
  .then((res) => {
    console.log(res.stdout)
  })
  .catch((err) => {
    console.log('failed', err)
  })

console.log('starting')
setTimeout(() => {
  console.log('exiting')
  process.exit()
}, 10 * 1000)
```

Running this will print `exiting` after `10s` and exit but when I grep for `ipfs dameon` the process is still running.

### PR DESCRIPTION
**Pull request by dignifiedquire**
As spawned processes are not cleaned up when the main process dies
this option adds the abilitiy to kill those when the spawning process
dies.

Closes #51

### CODE REVIEW

**Comment by dignifiedquire** [comment]
Looks like this is not working on CI :( passed fine for me locally will investigate what the issue is.

File: index.js [hunk 1]
```diff
@@ -33,7 +34,8 @@ function handleArgs(cmd, args, opts) {
 		stripEof: true,
 		preferLocal: true,
 		encoding: 'utf8',
-		reject: true
+		reject: true,
+		cleanup: false
```
Comment (sindresorhus) [inline index.js]: Any reason we wouldn't want this as `true` by default? I already assumed that's how the core one worked. We could clearly document it in the readme.

File: readme.md [hunk 2]
```diff
@@ -11,7 +11,7 @@
 - [Improved Windows support.](https://github.com/IndigoUnited/node-cross-spawn#why)
 - Higher max buffer. 10 MB instead of 200 KB.
 - [Executes locally installed binaries by name.](#preferlocal)
-
```
Comment (sindresorhus) [inline readme.md]: Don't remove the newline

File: readme.md [hunk 3]
```diff
@@ -11,7 +11,7 @@
...
+- [Cleans up spawned processes when the parent process dies](#cleanup)
```
Comment (sindresorhus) [inline readme.md]: Should end in a `.`

File: readme.md [hunk 4]
```diff
@@ -138,6 +138,12 @@ Default: `true`
 
 Setting this to `false` resolves the promise with the error instead of rejecting it.
 
+#### cleanup
+
+Type: `boolean`<br>
+Default: `false`
+
+Setting this to `true` keeps track of the spawned process and `kill`s it when the parent process exits.
```
Comment (sindresorhus) [inline readme.md]: > Keep track of the spawned process and `kill`s it when the parent process exits.

File: test.js [hunk 5]
```diff
@@ -276,3 +277,24 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
 test(cmd, ' foo bar', 'foo', 'bar');
 test(cmd, ' baz quz', 'baz', 'quz');
 test(cmd, '');
+
+async function spawnAndKill(t, signal) {
+	const cp = m('sub-process');
+	let pid;
+
```
Comment (sindresorhus) [inline test.js]: `cp.stdout.setEncoding('utf8')`

Then you don't have to `.toString()` it below.

File: test.js [hunk 6]
```diff
@@ -276,3 +277,24 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
...
+	cp.stdout.on('data', chunk => {
+		pid = parseInt(chunk.toString(), 10);
+		t.is(typeof pid, 'number');
+
+		setTimeout(function () {
```
Comment (sindresorhus) [inline test.js]: use arrow function

File: fixtures/sub-process [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env node
+'use strict';
+
```
Comment (sindresorhus) [inline fixtures/sub-process]: remove this newline

File: index.js (hunk 1 above)
Comment (dignifiedquire) [inline index.js]: I was just conscious of this being a breaking change and thought you might not want that, happy to default it to `true`.

Comment (sindresorhus) [inline index.js]: We already have other breaking changes in master. So will be a breaking release regardless.

Comment (dignifiedquire) [inline index.js]: Okay, will change then

**Comment by dignifiedquire** [comment]
@sindresorhus squashed commits and travis seems to be happy now that I introduced a small delay for the pid check.

I've also added a test to make sure it behaves as expected when using `cleanup: false`

File: test.js [hunk 8]
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
...
+
+		setTimeout(() => {
+			process.kill(cp.pid, signal);
+		}, 100);
+	});
+
+	await t.throws(cp);
+
+	// Give everybody some time to breath and kill things
+	const timeout = delay => new Promise(resolve => setTimeout(resolve, delay));
```
Comment (sindresorhus) [inline test.js]: Just use https://github.com/sindresorhus/delay

File: test.js [hunk 9]
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
...
+
+	await timeout(100);
```
Comment (sindresorhus) [inline test.js]: 100=>200 just to be safe

File: test.js [hunk 10]
```diff
@@ -276,3 +277,54 @@ cmd.title = (message, expected) => `cmd is: ${JSON.stringify(expected)}`;
...
+
+	cp.stdout.setEncoding('utf8');
+	cp.stdout.on('data', chunk => {
+		pid = parseInt(chunk, 10);
+		t.is(typeof pid, 'number');
+
+		setTimeout(() => {
+			process.kill(cp.pid, 'SIGINT');
+		}, 100);
+	});
```
Comment (sindresorhus) [inline test.js]: This is almost a duplicate of the other test. Can you extract this into a function both could use?

File: test.js (hunk 8 above)
Comment (dignifiedquire) [inline test.js]: I knew there was a module somewhere ;)

**Comment by dignifiedquire** [comment]
Squashed and green (at least before squash)

**Comment by dignifiedquire** [comment]
@sindresorhus ping

**Comment by sindresorhus** [comment]
Thank you! :)

Sorry about the delay. I've been quite busy.

Unfortunately we can't do a new release with this until https://github.com/sindresorhus/execa/issues/46 is resolved... Any help welcome.

**Comment by dignifiedquire** [comment]
> Sorry about the delay. I've been quite busy.

No need to be sorry :) Thank you for merging

**Pull request merged**